README
App-LedgerSMB-Admin

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::LedgerSMB::Admin

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-LedgerSMB-Admin

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-LedgerSMB-Admin

    CPAN Ratings
        http://cpanratings.perl.org/d/App-LedgerSMB-Admin

    Search CPAN
        http://search.cpan.org/dist/App-LedgerSMB-Admin/


LICENSE AND COPYRIGHT

Copyright (C) 2014 Chris Travers

This program is distributed under the (Revised) BSD License:
L<http://www.opensource.org/licenses/BSD-3-Clause>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of Chris Travers's Organization
nor the names of its contributors may be used to endorse or promote
products derived from this software without specific prior written
permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for App-LedgerSMB-Admin

0.06    2017-05-10
        Remove unused dependencies

0.04    2014-10-12
        Added stats function for databases.

0.03    2014-10-10
        Documentation fixes, streamlining of module, looking forward.

0.01    2014
        First version, released on an unsuspecting world.



Makefile.PL
use 5.010;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'App::LedgerSMB::Admin',
    AUTHOR           => q{Chris Travers <chris@efficito.com>},
    VERSION_FROM     => 'lib/App/LedgerSMB/Admin.pm',
    ABSTRACT_FROM    => 'lib/App/LedgerSMB/Admin.pm',
    LICENSE          => 'BSD',
    PL_FILES         => {},
    EXE_FILES        => [ 'bin/lsmb_reload', 'bin/lsmb_createdb' ],
    MIN_PERL_VERSION => 5.010,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
        "Test::Exception" => 0,
    },
    PREREQ_PM => {
        "PGObject::Util::DBMethod" => 0,
        "PGObject::Util::DBAdmin"  => 0,
        "PGObject::Util::DBChange"  => 0.04,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'App-LedgerSMB-Admin-*' },
    META_MERGE => {
       'meta-spec' => { version => 2 },
        resources => {
           repository => {
             type => 'git',
             url  => 'https://github.com/ledgersmb/App-LedgerSMB-Admin.git',
             web  => 'https://github.com/ledgersmb/App-LedgerSMB-Admin',
           },
       },
    },
);


lib/App/LedgerSMB/Admin.pm
package App::LedgerSMB::Admin;

use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

App::LedgerSMB::Admin - Easily Manage LedgerSMB Installations

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';


=head1 SYNOPSIS

This module provides the basic routines for managing multiple major versions of
LedgerSMB through a consistent toolkit.

It contains basically two components:

=over

=item command line scripts

=item libraries for writing management programs

=back

=head1 BUNDLED CLI PROGRAMS

=head2 lsmb_reload

Due to improper errors, this currently does not work properly with LedgerSMB
1.3.  It is expected that it will following the LedgerSMB 1.3.45 release
(forthcoming).

The lsmb_reload application provides a command line tool for upgrading or
rebuilding stored procedures and permissions of a LedgerSMB 1.3 or 1.4 database.

Due to shortcomings in error handling it does not work for LedgerSMB 1.3.44 and
below.

The program supports UNIX sockets as well as TCP/IP for connecting to the
LedgerSMB database, and all aspects of the database connection may be controlled
through the same environment variables as any other libpq program.

The following options are supported:

   --all                    Reload All Databases
   --help                   Print this message and exit
   --host hostname          Database Host
   --path13 /example/path   Path to LedgerSMB 1.3 installation
   --path14 /example/path2  Path to LedgerSMB 1.4 installation
   --port 5432              Database Poart
   --dbname database        Reload the specified db, overridden by --all
   --username postgres      Database Superuser to Log In As
   --prompt-password        Prompt for Password (can use PGPASSWORD instead)


=head2 lsmb_createdb

Due to improper errors, this currently does not work properly with LedgerSMB
1.3.  It is expected that it will following the LedgerSMB 1.3.45 release
(forthcoming).

The lsmb_reload application provides a command line tool for upgrading or
rebuilding stored procedures and permissions of a LedgerSMB 1.3 or 1.4 database.

   --help                   Print this message and exit
   --host hostname          Database Host
   --path /example/path     Path to LedgerSMB installation (required)
   --chart us/chart/General Chart of Accounts path (relative to sql)
   --gifi  ca/gifi/General  Path to GIFI
   --port 5432              Database Poart
   --dbname database        Create db with the following name (required)
   --username postgres      Database Superuser to Log In As
   --prompt-password        Prompt for Password (can use PGPASSWORD instead)

=head1 Bundled Libraries

=head2 App::LedgerSMB::Admin

Provides base version tracking and management routines

=head2 App::LedgerSMB::Admin::Database

Routines to reload, create, update, backup, and restore databases

=head2 App::LedgerSMB::Admin::Database::Setting

Look up settings in a LedgerSMB database

=head2 App::LedgerSMB::Admin::User

Undeveloped, but will include basic user management routines.

=head1 SUBROUTINES/METHODS of This Library

=head2 add_paths(%versionhash)

=cut

my %version_paths;

sub add_paths{
    shift if $_[0] eq __PACKAGE__;
    my %version_hash = @_;
    %version_paths = (%version_paths, %version_hash);
    return %version_paths;
}

=head2 path_for($major_version)

=cut

sub path_for{
    shift if $_[0] eq __PACKAGE__;
    return $version_paths{$_[0]};
}

=head1 AUTHOR

Chris Travers, C<< <chris at efficito.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-app-ledgersmb-admin at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=App-LedgerSMB-Admin>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::LedgerSMB::Admin


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-LedgerSMB-Admin>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/App-LedgerSMB-Admin>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/App-LedgerSMB-Admin>

=item * Search CPAN

L<http://search.cpan.org/dist/App-LedgerSMB-Admin/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 Chris Travers.

This program is distributed under the (Revised) BSD License:
L<http://www.opensource.org/licenses/BSD-3-Clause>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of Chris Travers's Organization
nor the names of its contributors may be used to endorse or promote
products derived from this software without specific prior written
permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of App::LedgerSMB::Admin


lib/App/LedgerSMB/Admin/Auth.pm
=head1 NAME

App::LedgerSMB::Admin::Auth - Authenticate Web Users for LSMB Admin Tasks

=cut

package App::LedgerSMB::Admin::Auth;
use Dancer ':syntax';
use Dancer::Response;
use HTTP::Headers;
use MIME::Base64;
use strict;
use warnings;

=head1 SYNOPSIS

 my $db = authenticate(host => 'localhost', port => 5432, dbname => 'lsmb');

=head1 FUNCTIONS

=head2 authenticate

This function grabs the parameters of host and port from the arg hash and the
username and password from the authentication header, and returns a db object
with a connection.  If the connection fails, it returns a 401 error.

Currently only basic auth is supported but there are plans to add krb5 auth
in the near future.

We currently return a 401 any time the db connection fails for any reason.
This is to prevent using the tool to scan for db servers for further attacks.

=cut

sub authenticate {
    my %args = @_;
    my $header = request->header('Authorization');
    my ($user, $password) = split(/:/, (MIME::Base64::decode($1) || ":"));
    return _fail_auth() unless $user;
    my $db = App::LedgerSMB::Admin::Database->new(%args,
                                                  username => $user,
                                                  password => $password,
    );
    return $db if $db->connect;
    return _fail_auth();
}

sub _fail_auth {
    my $authmsg = 'Authorization Required';    
    return halt(Dancer::Response->new(
        status => 401,
        content => $authmsg,
        headers => [
            'Content-Type' => 'text/plain',
            'Content-Length' => length($authmsg),
            'WWW-Authenticate' => 'Basic realm="LedgerSMB Admin"',
        ]
    ));
}

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Chris Travers.

This program is distributed under the (Revised) BSD License:
L<http://www.opensource.org/licenses/BSD-3-Clause>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of Chris Travers's Organization
nor the names of its contributors may be used to endorse or promote
products derived from this software without specific prior written
permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1;


lib/App/LedgerSMB/Admin/Database/Setting.pm
=head1 NAME

App::LedgerSMB::Admin::Database::Setting - Access LSMB Settings from Admin Tools

=cut

package App::LedgerSMB::Admin::Database::Setting;
use Moo;
with 'PGObject::Simple::Role';

use PGObject::Util::DBMethod;

sub _get_dbh {
    my ($self) = @_;
    return $self->database->connect;
}

sub _get_prefix { 'setting__' };

=head1 SYNOPSIS

 my $db = App::LedgerSMB::Admin::Database->new(
      username => 'postgres',
      password => 'secret',
      host     => 'localhost',
      port     => '5432',
 );
 my $setting = App::LedgerSMB::Admin::Database::Setting->new(
     database    => $db,
     setting_key => 'separate_duties');
 my $sep_duties = $setting->value;

=head1 PROPERTIES

=head2 database

The database for the associated setting.  Must be a 
App::LedgerSMB:Admin::Database object

=cut

has database => (
    is => 'ro', 
   isa => sub { die 'Must be a Database Object' 
                  unless eval { $_[0]->isa('App::LedgerSMB::Admin::Database')};
          }
);

=head2 setting_key

The name of the setting

=cut

has setting_key => (is => 'ro');

=head2 value

The value of the setting.  Is looked up lazily.

=cut

has value => (is => 'lazy');

sub _build_value {
    my $hash = $_[0]->_get;
    return $hash->{value};
}

dbmethod _get => (funcname => 'get');

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Chris Travers.

This program is distributed under the (Revised) BSD License:
L<http://www.opensource.org/licenses/BSD-3-Clause>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of Chris Travers's Organization
nor the names of its contributors may be used to endorse or promote
products derived from this software without specific prior written
permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1;


lib/App/LedgerSMB/Admin/Database/Routings.pm
package App::LedgerSMB::Admin::Database::Routings;
use Dancer ':syntax';
use Dancer::Serializer;
use Dancer::Plugin::Ajax;
use Template;

=head1 NAME

App::LedgerSMB::Admin::Database::Routings - Dancer App for LSMB DB Admin

=cut

=head1 URL LAYOUT

All urls occur in the /lsmbadmin/1.0/ path from the document root.

=cut

prefix '/lsmbadmin/1.0';

=head2 Per Server URLs

The server urls are then parameterized with host and port.

For each server the following resources are present:

=over

=item dbs - list databases (get only)

=item new - new database

=item globals - backup of server globals (put to restore)

=back

get retrieves information, post updates the information.

The complete urls then become, for a server at localhost and a port of 5432, 
off a base of:

 /lsmbadmin/1.0/localhost/5432/

So the database list would be found at

 /lsmbadmin/1.0/localhost/5432/dbs

=cut

get  '/:host/:port/dbs'     => sub { template 'dblist'  => _list_dbs() };
ajax  '/:host/:port/dbs'    => sub { to_json(_lsmb_dbs()) };
get  '/:host/:port/new'     => sub { template 'new_db'  => {} };
post '/:host/:port/new'     => sub { template 'db_info' => _createdb() };
ajax  '/:host/:port/new'    => sub { to_json(createdb()) };
get  '/:host/:port/globals' => sub { _backup_globals() };
put  '/:host/:port/globals' => sub { template 'restored' => _restore_globals()};
ajax '/:host/:port/globals' => sub { to_json(_restore_globals()) };

=head2 Per Database URLs

=cut

get  '/:host/:port/db/:dbname/' => sub { template 'db_info' => _db_info(); };
post '/:host/:port/db/:dbname/' => \&_reload_db;
put  '/:host/:port/db/:dbname/' => \&_reload_db;
put  '/:host/:port/db/:dbname/*.sql' => \&_run_file;
post '/:host/:port/db/:dbname/*.sql' => \&_run_file;
get  '/:host/:port/db/:dbname/backup/*.*' => \&_backup_db;
put  '/:host/:port/db/:dbname/backup/*.*' => \&_restore_backup;
post '/:host/:port/db/:dbname/backup/*.*' => \&_restore_backup;


sub _list_dbs {
    my $db = authenticate(
              host   => param('host'), 
              port   => param('port'),
              dbname => 'postgres'
    );
    return $db->list_dbs;
}

sub _createdb {
    my $db = authenticate(
              host   => param('host'), 
              port   => param('port'),
              dbname => 'postgres'
    );
    my $newdb = App::LedgerSMB::Admin::Database->new(
           $db->export, dbname => param('dbname')
    );
    $newdb->create;
    $newdb->load;
    return $newdb->list_dbs;
}

sub _backup_globals {
}

sub _restore_globals {
}

sub _reload_db {
}

sub _backup_db {
}

sub _restore_db {
}

sub _run_file {
}

1;


lib/App/LedgerSMB/Admin/Database.pm
package App::LedgerSMB::Admin::Database;
use Moo;
extends 'PGObject::Util::DBAdmin';
use File::Temp;
use Cwd;
use App::LedgerSMB::Admin;
use App::LedgerSMB::Admin::Database::Setting;

=head1 NAME

App::LedgerSMB::Admin::Database - Administer LedgerSMB Databases

=head1 SYNOPSIS

Upgrading to 1.4 from 1.3, after updating 1.3 instance to latest:

  use App::LedgerSMB::Admin;
  use App::LedgerSMB::Admin::Database;
  App::LedgerSMB::Admin->add_paths(
     '1.3' => '/usr/local/ledgersmb_1.3',
     '1.4' => '/usr/local/ledgersmb_1.4',
  ); # setting the version paths
  my $db = App::LedgerSMB::Admin::Database->new(
     username  => 'postgres',
     password  => 'secretpassword',
     host      => 'localhost',
     port      => '5432',
     dbname    => 'mycompany',
  );
  if ($db->major_version eq '1.3') {
      $db->reload;
      $db->upgrade_to('1.4');
  }

=head1 VERSION

0.04

=cut

our $VERSION=0.04;

=head1 PROPERTIES INHERITED FROM PGObject::Util::DBAdmin

Please see the docs for PGObject::Util::DBAdmin.

=head2 username 

=head2 password

=head2 host

=head2 port

=head2 dbname

=head1 ADDITIONAL PROPERTIES

=head2 version

Returns the version number of the database.

=cut

has version => (is => 'lazy');

sub _build_version {
    my $self = shift;
    return App::LedgerSMB::Admin::Database::Setting->new(
                                        database    => $self,
                                        setting_key => 'version')->value;
}

=head2 major_version

Major versions are generally understood to be not backwards compatible.  In
LedgerSMB, as with PostgreSQL, major versions are based on the second numbers
in the version, so 1.2, 1.3, and 1.4 are major versions.

=cut

has major_version => (is => 'lazy');

sub _build_major_version {
    my $self = shift;
    my $version = $self->version;
    $version =~ s/\.\d*(?:-dev)?$//;
    return $version;
}

=head1 METHODS INHERITED

Please see the docs for PGObject::Util::DBAdmin

=head2 create

=head2 connect

=head2 drop

=head2 run_file

=head2 backup

=head2 restore_backup

=head1 NEW METHODS

=head2 stats

Returns a hashref of table names to rows.  The following tables are counted:

=over

=item ar

=item ap

=item gl

=item oe

=item acc_trans

=item users

=item entity_credit_account

=item entity

=back

=cut

my @tables = qw(ar ap gl users entity_credit_account entity acc_trans oe);

sub stats {
    my ($self) = @_;
    my $dbh = $self->connect;
    my $results;

    $results->{$_->{table}} = $_->{count}
    for map {
       my $sth = $dbh->prepare($_->{query});
       $sth->execute;
       my ($count) = $sth->fetchrow_array;
       { table => $_->{table}, count => $count };
    } map {
       my $qt = 'SELECT COUNT(*) FROM __TABLE__';
       my $id = $dbh->quote_identifier($_);
       $qt =~ s/__TABLE__/$id/;
       { table => $_, query => $qt };
    } @tables;

    return $results;
}

=head2 load($major_version)

Loads the db schema for the major version requested.

=cut

sub load {
    my ($self, $major_version) = @_;
    eval {
       $self->run_file(
            file => App::LedgerSMB::Admin->path_for($major_version)
                    . "/sql/Pg-database.sql"
       );
    };
    my $sqlpath = App::LedgerSMB::Admin->path_for($major_version) . "/sql/modules";
    return $self->process_loadorder($sqlpath, "$sqlpath/LOADORDER");
}

=head2 reload

Reloads all modules in a LedgerSMB instance.

=cut

sub reload {
    my ($self) = @_;
    my $path = Cwd::getcwd();
    my $sqlpath = App::LedgerSMB::Admin->path_for($self->major_version);
    chdir $sqlpath;
    my $rc =  $self->process_loadorder('sql/modules', "sql/modules/LOADORDER");
    chdir $path;
    return $rc;
}

=head2 process_loadorder($sql_path, $loadorder_path);

Processes a specific loadorder.  Useful for installing extensions to LedgerSMB

Dies if any SQL files produce errors except from a file starting with "Fixes."

=cut

sub process_loadorder {
    my ($self, $sql_path, $loadorder_path) = @_;
    $sql_path =~ s|/$||;
    for my $line (_loadorder_entries($loadorder_path)){
        if ($line =~ /^Fixes/){
           eval { $self->run_file(file => "$sql_path/$line") };
        } else {
           $self->run_file(file => "$sql_path/$line");
        }
    }
    $self->process_old_roles() if $self->major_version eq '1.3';
    return 1;
}

sub _loadorder_entries {
    my $loadorderpath = shift;
    open(LOAD, '<', $loadorderpath) || die "Cannot open loadorder: $!";
    return grep {$_} map { my $l = $_; $l =~ s/(\s*|#.*)//g; $l} <LOAD>;
}

=head2 process_changes($loadorderfile)
applies db changes (post-1.4) to the db as specified in the provided LOADORDER

=cut

sub process_changes {
    my ($self, $loadorderpath) = @_;
    my $dbh = $self->connect;
    my $sql_path = $loadorderpath;
    $sql_path =~ s#/?LOADORDER$##;
    PGObject::Util::DBChange->init_if_needed($dbh);
    for my $line (_loadorder_entries($loadorderpath)){
        my $line =~ s/^(!)?//;
        my $failure_ok = $1;
        my $dbchange = PGObject::Util::DBChange->new(
                     path => "$sql_path/$line",
                     no_transactions => 1
        ); 
        $dbchange->apply($dbh) || $failure_ok 
           || die "Change $line failed: " . $dbh->ERRSTR;
    }
}

=head2 process_old_roles($rolefile)

Processes an old-style (1.3-era) roles file and runs it on the database.

=cut

sub process_old_roles {
    my ($self, $rolefile, %args) = @_;
    $rolefile ||= 'Roles.sql';
    my $sqlpath = App::LedgerSMB::Admin->path_for($self->major_version)
                  . '/sql/modules';
    my $tempdir = $args{tempdir} || $ENV{TEMP} || '/tmp';
    my $temp = File::Temp->new(DIR => $tempdir);
    open ROLES, '<', "$sqlpath/$rolefile";
    for my $line (<ROLES>){
        my $dbname = $self->dbname;
        $line =~ s/<\?lsmb dbname ?>/$dbname/g;
        print $temp $line;
    }
    eval { $self->run_file($temp->filename); };
}

=head2 upgrade_to($major_version)

Upgrades to the major version specified.  Provides an error if the upgrade file
is not found.

=cut

sub upgrade_to {
    my ($self, $major_version) = @_;
    my $up_filename = "lsmb" . $self->major_version . '-' . $major_version
                      . ".sql";
    my $versionpath = App::LedgerSMB::Admin->path_for($major_version);
    die 'No version path registered' unless $versionpath;
    $self->run_file(file => "$versionpath/sql/upgrade/$up_filename");
    $self->new($self->export)->reload;
}

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Chris Travers.

This program is distributed under the (Revised) BSD License:
L<http://www.opensource.org/licenses/BSD-3-Clause>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of Chris Travers's Organization
nor the names of its contributors may be used to endorse or promote
products derived from this software without specific prior written
permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1;	;
